File: src/wann/mutations.py
"""WANN mutation operators.

Available mutations:
  - Add node: split an existing enabled connection, insert hidden node.
  - Add connection: create a new connection between two unconnected nodes.
  - Toggle connection: enable/disable a random connection.
  - Flip sign: flip the sign of a random connection.
  - Change activation: re-roll activation function of a random node.
  - Change aggregation: re-roll aggregation function of a random node.

No weight mutation — weights are sign-only and shared across the network.
"""

from __future__ import annotations

import numpy as np

from src.wann.genome import (
    FIRST_HIDDEN_ID,
    OUTPUT_START,
    BIAS_ID,
    ConnGene,
    Genome,
    NodeGene,
    NodeType,
)
from src.wann.logical_nodes import (
    ALL_ACTIVATIONS,
    ALL_AGGREGATIONS,
    ActivationFn,
    AggregationFn,
)


def _pick_random_connection(
    genome: Genome, rng: np.random.Generator
) -> ConnGene | None:
    """Pick a random connection from the genome."""
    conns = list(genome.conn_genes.values())
    if not conns:
        return None
    return conns[rng.integers(len(conns))]


def _pick_random_enabled_connection(
    genome: Genome, rng: np.random.Generator
) -> ConnGene | None:
    """Pick a random enabled connection."""
    enabled = [c for c in genome.conn_genes.values() if c.enabled]
    if not enabled:
        return None
    return enabled[rng.integers(len(enabled))]


def _pick_random_node(
    genome: Genome,
    rng: np.random.Generator,
    exclude_inputs: bool = True,
    exclude_bias: bool = True,
    exclude_outputs: bool = True,
) -> NodeGene | None:
    """Pick a random node gene."""
    candidates = list(genome.node_genes.values())
    if exclude_inputs:
        candidates = [n for n in candidates if n.node_type != NodeType.INPUT]
    if exclude_bias:
        candidates = [n for n in candidates if n.node_type != NodeType.BIAS]
    if exclude_outputs:
        candidates = [n for n in candidates if n.node_type != NodeType.OUTPUT]
    if not candidates:
        return None
    return candidates[rng.integers(len(candidates))]


def mutate_add_node(genome: Genome, rng: np.random.Generator) -> bool:
    """Split a random enabled connection, inserting a new hidden node.

    The original connection (src→dst) is disabled. Two new connections are
    added: src→new_node (sign=+1) and new_node→dst (sign=original_sign).
    The new node uses IDENTITY activation and SUM aggregation.
    """
    conn = _pick_random_enabled_connection(genome, rng)
    if conn is None:
        return False

    # Disable the old connection.
    genome.conn_genes[conn.innovation] = ConnGene.make(
        conn.innovation, conn.src, conn.dst, conn.sign, enabled=False
    )
    genome.invalidate_topo_order()

    # Create a new hidden node.
    new_id = genome.node_ids
    new_id_max = max(new_id) if new_id else FIRST_HIDDEN_ID - 1
    node_id = max(new_id_max + 1, FIRST_HIDDEN_ID)
    new_node = NodeGene.make(
        node_id, NodeType.HIDDEN, ActivationFn.IDENTITY, AggregationFn.SUM
    )
    genome.add_node(new_node)

    # Add src → new_node (sign=+1).
    inno1 = genome.next_innovation
    genome.add_connection(ConnGene.make(inno1, conn.src, node_id, sign=1))
    genome.next_innovation = max(genome.next_innovation, inno1 + 1)

    # Add new_node → dst (sign=original_sign).
    inno2 = genome.next_innovation
    genome.add_connection(ConnGene.make(inno2, node_id, conn.dst, sign=conn.sign))
    genome.next_innovation = max(genome.next_innovation, inno2 + 1)

    return True


def mutate_add_connection(genome: Genome, rng: np.random.Generator) -> bool:
    """Add a new connection between two unconnected nodes.

    Avoids creating cycles by only allowing connections from lower-depth
    nodes to higher-depth nodes (or same depth for recurrent — but we
    keep it feed-forward for simplicity).
    """
    order = genome.topological_order()
    depth = {nid: i for i, nid in enumerate(order)}

    # Candidate pairs: src must come before dst in topological order.
    candidates: list[tuple[int, int]] = []
    existing = {(c.src, c.dst) for c in genome.conn_genes.values()}

    for src in order:
        for dst in order:
            if depth[src] >= depth[dst]:
                continue
            if (src, dst) in existing:
                continue
            # Don't connect output->anything.
            src_node = genome.node_genes.get(src)
            if src_node is not None and src_node.node_type == NodeType.OUTPUT:
                continue
            # Don't connect anything->input or anything->bias.
            dst_node = genome.node_genes.get(dst)
            if dst_node is not None and dst_node.node_type in (
                NodeType.INPUT,
                NodeType.BIAS,
            ):
                continue
            candidates.append((src, dst))

    if not candidates:
        return False

    src, dst = candidates[rng.integers(len(candidates))]
    sign = rng.choice([-1, 1])
    inno = genome.next_innovation
    genome.add_connection(ConnGene.make(inno, src, dst, sign=sign))
    genome.next_innovation = max(genome.next_innovation, inno + 1)
    return True


def mutate_toggle_connection(genome: Genome, rng: np.random.Generator) -> bool:
    """Enable or disable a random connection."""
    conn = _pick_random_connection(genome, rng)
    if conn is None:
        return False
    genome.conn_genes[conn.innovation] = ConnGene.make(
        conn.innovation, conn.src, conn.dst, conn.sign, enabled=not conn.enabled
    )
    genome.invalidate_topo_order()
    return True


def mutate_flip_sign(genome: Genome, rng: np.random.Generator) -> bool:
    """Flip the sign of a random connection (+1 → -1 or vice versa)."""
    conn = _pick_random_connection(genome, rng)
    if conn is None:
        return False
    new_sign = -conn.sign
    genome.conn_genes[conn.innovation] = ConnGene.make(
        conn.innovation, conn.src, conn.dst, sign=new_sign, enabled=conn.enabled
    )
    genome.invalidate_topo_order()
    return True


def mutate_change_activation(genome: Genome, rng: np.random.Generator) -> bool:
    """Change the activation function of a random non-input node."""
    node = _pick_random_node(genome, rng)
    if node is None:
        return False
    current = node.activation_fn
    choices = [a for a in ALL_ACTIVATIONS if a != current]
    new_fn = choices[rng.integers(len(choices))]
    genome.node_genes[node.id] = NodeGene.make(
        node.id, node.node_type, new_fn, node.aggregation_fn
    )
    return True


def mutate_change_aggregation(genome: Genome, rng: np.random.Generator) -> bool:
    """Change the aggregation function of a random non-input node."""
    node = _pick_random_node(genome, rng)
    if node is None:
        return False
    current = node.aggregation_fn
    choices = [a for a in ALL_AGGREGATIONS if a != current]
    new_fn = choices[rng.integers(len(choices))]
    genome.node_genes[node.id] = NodeGene.make(
        node.id, node.node_type, node.activation_fn, new_fn
    )
    return True


# All mutation operators for sampling.
MUTATIONS = [
    mutate_add_node,
    mutate_add_connection,
    mutate_toggle_connection,
    mutate_flip_sign,
    mutate_change_activation,
    mutate_change_aggregation,
]

File: src/sueca_wann/src/mutations.rs
use crate::genome::{
    ActivationFn, AggregationFn, ConnGene, Genome, NodeGene, NodeType, FIRST_HIDDEN_ID,
};
use rand::Rng;
use std::collections::{HashMap, HashSet};

pub struct InnovationRegistry {
    pub next_innovation: usize,
    pub conns: HashMap<(usize, usize), usize>,
}

impl InnovationRegistry {
    pub fn new(next_innovation: usize) -> Self {
        Self {
            next_innovation,
            conns: HashMap::new(),
        }
    }

    pub fn get_or_create(&mut self, src: usize, dst: usize) -> usize {
        if let Some(&inno) = self.conns.get(&(src, dst)) {
            inno
        } else {
            let inno = self.next_innovation;
            self.conns.insert((src, dst), inno);
            self.next_innovation += 1;
            inno
        }
    }
}

fn pick_random_connection<R: Rng>(genome: &Genome, rng: &mut R) -> Option<ConnGene> {
    let conns: Vec<ConnGene> = genome.conn_genes.values().cloned().collect();
    if conns.is_empty() {
        return None;
    }
    let idx = rng.gen_range(0..conns.len());
    Some(conns[idx].clone())
}

fn pick_random_enabled_connection<R: Rng>(genome: &Genome, rng: &mut R) -> Option<ConnGene> {
    let enabled: Vec<ConnGene> = genome
        .conn_genes
        .values()
        .filter(|c| c.enabled)
        .cloned()
        .collect();
    if enabled.is_empty() {
        return None;
    }
    let idx = rng.gen_range(0..enabled.len());
    Some(enabled[idx].clone())
}

fn pick_random_node<R: Rng>(
    genome: &Genome,
    rng: &mut R,
    exclude_inputs: bool,
    exclude_bias: bool,
    exclude_outputs: bool,
) -> Option<NodeGene> {
    let mut candidates: Vec<NodeGene> = genome.node_genes.values().cloned().collect();
    if exclude_inputs {
        candidates.retain(|n| n.node_type != NodeType::INPUT);
    }
    if exclude_bias {
        candidates.retain(|n| n.node_type != NodeType::BIAS);
    }
    if exclude_outputs {
        candidates.retain(|n| n.node_type != NodeType::OUTPUT);
    }
    if candidates.is_empty() {
        return None;
    }
    let idx = rng.gen_range(0..candidates.len());
    Some(candidates[idx].clone())
}

pub fn mutate_add_node<R: Rng>(
    genome: &mut Genome,
    registry: &mut InnovationRegistry,
    rng: &mut R,
) -> bool {
    let conn = match pick_random_enabled_connection(genome, rng) {
        Some(c) => c,
        None => return false,
    };

    // Disable the old connection
    if let Some(c) = genome.conn_genes.get_mut(&conn.innovation) {
        c.enabled = false;
    }

    // Create a new hidden node
    let existing_ids = genome.node_ids();
    let max_id = existing_ids
        .iter()
        .max()
        .copied()
        .unwrap_or(FIRST_HIDDEN_ID - 1);
    let node_id = max_id + 1;

    let all_activations = [
        ActivationFn::IDENTITY,
        ActivationFn::NOT,
        ActivationFn::THRESHOLD,
    ];
    let all_aggregations = [AggregationFn::SUM, AggregationFn::MIN, AggregationFn::MAX];

    let act = all_activations[rng.gen_range(0..all_activations.len())];
    let agg = all_aggregations[rng.gen_range(0..all_aggregations.len())];

    let new_node = NodeGene::make(node_id, NodeType::HIDDEN, act, agg);
    genome.add_node(new_node);

    // Add src -> new_node (sign = +1)
    let inno1 = registry.get_or_create(conn.src, node_id);
    genome.add_connection(ConnGene::make(inno1, conn.src, node_id, 1, true));

    // Add new_node -> dst (sign = original_sign)
    let inno2 = registry.get_or_create(node_id, conn.dst);
    genome.add_connection(ConnGene::make(inno2, node_id, conn.dst, conn.sign, true));

    genome.next_innovation = registry.next_innovation;

    true
}

pub fn mutate_add_connection<R: Rng>(
    genome: &mut Genome,
    registry: &mut InnovationRegistry,
    rng: &mut R,
) -> bool {
    let order = genome.topological_order();
    let mut depth = HashMap::new();
    for (i, &nid) in order.iter().enumerate() {
        depth.insert(nid, i);
    }

    let existing: HashSet<(usize, usize)> =
        genome.conn_genes.values().map(|c| (c.src, c.dst)).collect();

    let mut sources = Vec::new();
    let mut destinations = Vec::new();

    for &nid in &order {
        if let Some(node) = genome.node_genes.get(&nid) {
            if node.node_type != NodeType::OUTPUT {
                sources.push(nid);
            }
            if node.node_type != NodeType::INPUT && node.node_type != NodeType::BIAS {
                destinations.push(nid);
            }
        }
    }

    let mut candidates = Vec::new();
    for &src in &sources {
        let src_depth = depth[&src];
        for &dst in &destinations {
            if src_depth >= depth[&dst] {
                continue;
            }
            if !existing.contains(&(src, dst)) {
                candidates.push((src, dst));
            }
        }
    }

    if candidates.is_empty() {
        return false;
    }

    let idx = rng.gen_range(0..candidates.len());
    let (src, dst) = candidates[idx];
    let sign = if rng.gen_bool(0.5) { 1 } else { -1 };

    let inno = registry.get_or_create(src, dst);
    genome.add_connection(ConnGene::make(inno, src, dst, sign, true));

    genome.next_innovation = registry.next_innovation;

    true
}

pub fn mutate_toggle_connection<R: Rng>(genome: &mut Genome, rng: &mut R) -> bool {
    let conn = match pick_random_connection(genome, rng) {
        Some(c) => c,
        None => return false,
    };

    if let Some(c) = genome.conn_genes.get_mut(&conn.innovation) {
        c.enabled = !c.enabled;
        true
    } else {
        false
    }
}

pub fn mutate_flip_sign<R: Rng>(genome: &mut Genome, rng: &mut R) -> bool {
    let conn = match pick_random_connection(genome, rng) {
        Some(c) => c,
        None => return false,
    };

    if let Some(c) = genome.conn_genes.get_mut(&conn.innovation) {
        c.sign = -c.sign;
        true
    } else {
        false
    }
}

pub fn mutate_change_activation<R: Rng>(genome: &mut Genome, rng: &mut R) -> bool {
    let node = match pick_random_node(genome, rng, true, true, true) {
        Some(n) => n,
        None => return false,
    };

    let all_activations = [
        ActivationFn::IDENTITY,
        ActivationFn::NOT,
        ActivationFn::THRESHOLD,
    ];
    let current = node.activation_fn;
    let choices: Vec<ActivationFn> = all_activations
        .iter()
        .copied()
        .filter(|&a| a != current)
        .collect();
    if choices.is_empty() {
        return false;
    }

    let new_fn = choices[rng.gen_range(0..choices.len())];
    if let Some(n) = genome.node_genes.get_mut(&node.id) {
        n.activation_fn = new_fn;
        true
    } else {
        false
    }
}

pub fn mutate_change_aggregation<R: Rng>(genome: &mut Genome, rng: &mut R) -> bool {
    let node = match pick_random_node(genome, rng, true, true, true) {
        Some(n) => n,
        None => return false,
    };

    let all_aggregations = [AggregationFn::SUM, AggregationFn::MIN, AggregationFn::MAX];
    let current = node.aggregation_fn;
    let choices: Vec<AggregationFn> = all_aggregations
        .iter()
        .copied()
        .filter(|&a| a != current)
        .collect();
    if choices.is_empty() {
        return false;
    }

    let new_fn = choices[rng.gen_range(0..choices.len())];
    if let Some(n) = genome.node_genes.get_mut(&node.id) {
        n.aggregation_fn = new_fn;
        true
    } else {
        false
    }
}

pub fn apply_mutations<R: Rng>(
    genome: &mut Genome,
    registry: &mut InnovationRegistry,
    rng: &mut R,
    p_add_node: f64,
    p_add_conn: f64,
    p_toggle_conn: f64,
    p_flip_sign: f64,
    p_change_act: f64,
    p_change_agg: f64,
) -> usize {
    let mut n = 0;
    if rng.gen_bool(p_add_node) {
        if mutate_add_node(genome, registry, rng) {
            n += 1;
        }
    }
    if rng.gen_bool(p_add_conn) {
        if mutate_add_connection(genome, registry, rng) {
            n += 1;
        }
    }
    if rng.gen_bool(p_toggle_conn) {
        if mutate_toggle_connection(genome, rng) {
            n += 1;
        }
    }
    if rng.gen_bool(p_flip_sign) {
        if mutate_flip_sign(genome, rng) {
            n += 1;
        }
    }
    if rng.gen_bool(p_change_act) {
        if mutate_change_activation(genome, rng) {
            n += 1;
        }
    }
    if rng.gen_bool(p_change_agg) {
        if mutate_change_aggregation(genome, rng) {
            n += 1;
        }
    }
    n
}
